### ISSUE TITLE
LDAP-122: IllegalStateException: Cannot deactivate transaction synchronization - not active" thrown when when 'inner' transaction for LDAP context source fails to get a connection

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-ldap#160]
["Jacqueline Stark":https://jira.spring.io/secure/ViewProfile.jspa?name=jacqueline](Migrated from ["LDAP-122":https://jira.spring.io/browse/LDAP-122?redirect=false]) said:

Our project is using Spring-2.0.7 and Spring-ldap-1.2.1.

We have a HibernateTransactionManager on an outer service, calling an inner service, which has wrapped an LDAP ContextSourceTransactionManager.

If a transaction is attempted when LDAP isn't running, an exception occurs in Spring

java.lang.IllegalStateException: Cannot deactivate transaction synchronization - not active
 at org.springframework.transaction.support.TransactionSynchronizationManager.clearSynchronization(TransactionSynchronizationManager.java:274)
 at org.springframework.transaction.support.TransactionSynchronizationManager.clear(TransactionSynchronizationManager.java:412)
 at org.springframework.transaction.support.AbstractPlatformTransactionManager.cleanupAfterCompletion(AbstractPlatformTransactionManager.java:916)
 at org.springframework.transaction.support.AbstractPlatformTransactionManager.processRollback(AbstractPlatformTransactionManager.java:785)
 at org.springframework.transaction.support.AbstractPlatformTransactionManager.rollback(AbstractPlatformTransactionManager.java:730)

This then stops any future Hibernate transactions.

I found an issue against Spring core, http://jira.springframework.org/browse/SPR-3438 , which I think is the same problem.  This issue has been fixed in the extensions of AbstractPlatformTransactionManager in Spring core (eg HibernateTransactionManager, DataSourceTransactionManager).

The issue is resolved by catching Exception in the various implementations of doBegin(), and rethrowing a CannotCreateTransactionException, which is caught by AbstractPlatformTransactionManager.getTransaction(), which in turn resumes the outer transaction.

Should AbstractCompensatingTransactionManagerDelegate.doBegin() be doing the same? 

In other words, should the fix applied to Spring core for this issue also be applied to spring-ldap?

To reproduce, the same test code in http://jira.springframework.org/browse/SPR-3438 can be used, but with the txMgrInner being a ContextSourceTransactionManager.
